lib/bin/check-file-headers.js

```diff
@@ -19,50 +19,52 @@ const header = `// Copyright 2017 ODK Central Developers
 // except according to the terms contained in the LICENSE file.
 `;
 
-// This might not be the full list in local development: .gitignore includes
-// other files. However, this list should be sufficient for CircleCI.
-const skip = new Set([
-  '.circleci',
-  '.git',
-  '.github',
-  'config',
-  'docs',
+const skip = [
+  'test/',
   'lib/bin/.eslintrc.js',
   'lib/util/quarantine/pkcs7.js',
-  'node_modules',
-  'test',
-  '.eslintrc.json',
-  '.gitignore',
-  '.npmrc',
-  'CONTRIBUTING.md',
-  'LICENSE',
-  'Makefile',
-  'NOTICE',
-  'README.md',
-  'package-lock.json',
-  'package.json'
-]);
+];
 
-let missing = false;
-const checkDir = (dirPath) => {
-  for (const entry of fs.readdirSync(dirPath, { withFileTypes: true })) {
-    const entryPath = dirPath === '.' ? entry.name : `${dirPath}/${entry.name}`;
-    if (!skip.has(entryPath)) {
-      if (entry.isDirectory()) {
-        checkDir(entryPath);
-      } else {
-        const contents = fs.readFileSync(entryPath).toString();
-        const withConsistentYear = contents.replace(/(Copyright 20)\d\d/, '$117');
-        if (!withConsistentYear.startsWith(header)) {
-          console.error(entryPath);
-          missing = true;
-        }
+let stdin = '';
+process.stdin.resume();
+process.stdin.on('data', (data) => {
+  stdin += data.toString();
+});
+process.stdin.on('end', () => {
+  const files = stdin
+    .split('\n')
+    .filter(f => f.endsWith('.js'))
+    .filter(f => !skip.some(skipEntry => {
+      if (skipEntry.endsWith('.js')) return f === skipEntry;
+      if (skipEntry.endsWith('/')) return f.startsWith(skipEntry);
+      throw new Error(`Invalid skip entry: '${skipEntry}'`);
+    }));
+
+  const mode = process.argv[2] === '--fix' ? 'fix' : 'check';
+
+  let missing = false;
+
+  for (const entryPath of files) {
+    const contents = fs.readFileSync(entryPath).toString();
+    const withConsistentYear = contents.replace(/(Copyright 20)\d\d/, '$117');
```

Comment:
I've thought a little about this, and I think that might be tricky. We could try checking that all new files specify the current year in their file header — that'd work in most cases. But files are sometimes moved, which makes things hard: a file that was moved may look like a new file, yet specify a previous year in its file header. Things also get challenging around the year boundary, because a single PR may have commits from both years.

So I don't think CircleCI should fail if the year in the file header isn't the current year. If there's an easy way for it just to provide a warning in that case, I could see that being helpful in a follow-up PR — not sure what the possibilities there are.